pkg/lib/authn/authenticator/password/provider.go

```diff
@@ -159,9 +161,17 @@ func (p *Provider) Authenticate(a *authenticator.Password, password string) (ver
 func (p *Provider) UpdatePassword(a *authenticator.Password) error {
 	now := p.Clock.NowUTC()
 	a.UpdatedAt = now
-	a.ExpireAfter = nil
 
-	err := p.Store.UpdatePasswordHash(a)
+	prev, err := p.Store.Get(a.UserID, a.ID)
+	if err != nil {
+		return err
+	}
+
+	if prev.ExpireAfter == a.ExpireAfter {
+		a.ExpireAfter = nil
+	}
```

Comment:
The baseline is do not fetch the row again. Maybe you need to distinguish between the end-user update password (thus expire_after must be set to NULL), and Admin API change password (thus expire_after may be kept unchanged, or set to a value).